# GSE114944

[GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE114944)

[Reference](https://www.nature.com/articles/s41467-018-05072-0)

## General information
Identifying tumor antigen-specific T cells from cancer patients has been a goal of tumor immunologists for several decades. Co-expression of CD103 and CD39 on CD8 TIL highly enriches for tumor-reactive T cells. **CD39+CD103+ CD8 TIL** have a distinct TCR repertoire compared to other CD8 TIL subsets and are clonally expanded within the tumor. They are highly enriched for tumor antigen recognition and efficiently killed autologous tumor cells. Patients with head and neck cancer whose CD8 TIL contained a higher frequency of CD39+CD103+ cells experienced a greater overall survival. This work describes a simple and effective method for detecting tumor-reactive CD8 TIL.

## Sample information
Peripheral blood, uninvolved LNs, metastatic LNs, and tumor samples were obtained from  individuals with **head and neck sqaumous cell carcinoma (HNSCC), melanoma, colon cancer, rectal cancer, lung cancer, colorectal liver metastasis (CRLM), and ovarian cancer**. PBMCs were purified. Patients had no active treatment when the samples were collected. Previous treatment include surgeries, chemotherapies, immunotherapies and radiotherapies. 15 samples in total, each defined groups consists of 5 samples.

| Cell sorting |  |
| --- | --- |
| Naive CD8+ CTL | CD8+CD4−CD45RA+CCR7+ |
| Memory CD8+ CTL | CD8+CD4−CD45RA−CCR7+/− |
| DP CD8+ TIL | CD3+CD4−CD8+CD45RA−CCR7+/−**CD39+CD103+** |
| DN CD8+ TIL | CD3+CD4−CD8+CD45RA−CCR7+/−**CD39−CD103−** |
| CD103SP CD8+ TIL | CD3+CD4−CD8+CD45RA−CCR7+/−**CD39−CD103+** |

## GEO2R Code

Defined groups: DP, DN, CD103SP.

```r
library(GEOquery)
library(limma)
library(umap)

# load series and platform data from GEO
gset <- getGEO("GSE114944", GSEMatrix =TRUE, AnnotGPL=FALSE)
if (length(gset) > 1) idx <- grep("GPL15207", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]

# make proper column names to match toptable 
fvarLabels(gset) <- make.names(fvarLabels(gset))

# group membership for all samples
gsms <- "120100221020112"
sml <- strsplit(gsms, split="")[[1]]

# log2 transformation
ex <- exprs(gset)
qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))
LogC <- (qx[5] > 100) ||
          (qx[6]-qx[1] > 50 && qx[2] > 0)
if (LogC) { ex[which(ex <= 0)] <- NaN
  exprs(gset) <- log2(ex) }

# assign samples to groups and set up design matrix
gs <- factor(sml)
groups <- make.names(c("DP","DN","CD103SP"))
levels(gs) <- groups
gset$group <- gs
design <- model.matrix(~group + 0, gset)
colnames(design) <- levels(gs)

fit <- lmFit(gset, design)  # fit linear model

# set up contrasts of interest and recalculate model coefficients
cts <- paste(groups, c(tail(groups, -1), head(groups, 1)), sep="-")
cont.matrix <- makeContrasts(contrasts=cts, levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)

# compute statistics and table of top significant genes
fit2 <- eBayes(fit2, 0.01)
tT <- topTable(fit2, adjust="fdr", sort.by="B", number=250)

tT <- subset(tT, select=c("ID","adj.P.Val","P.Value","F","GB_ACC","GI","Gene.Symbol","SPOT_ID"))
write.table(tT, file=stdout(), row.names=F, sep="\t")

# Visualize and quality control test results.
# Build histogram of P-values for all genes. Normal test
# assumption is that most genes are not differentially expressed.
tT2 <- topTable(fit2, adjust="fdr", sort.by="B", number=Inf)
hist(tT2$adj.P.Val, col = "grey", border = "white", xlab = "P-adj",
  ylab = "Number of genes", main = "P-adj value distribution")

# summarize test results as "up", "down" or "not expressed"
dT <- decideTests(fit2, adjust.method="fdr", p.value=0.05)

# Venn diagram of results
vennDiagram(dT, circle.col=palette())

# create Q-Q plot for t-statistic
t.good <- which(!is.na(fit2$F)) # filter out bad probes
qqt(fit2$t[t.good], fit2$df.total[t.good], main="Moderated t statistic")

# volcano plot (log P-value vs log fold change)
colnames(fit2) # list contrast names
ct <- 1        # choose contrast of interest
volcanoplot(fit2, coef=ct, main=colnames(fit2)[ct], pch=20,
  highlight=length(which(dT[,ct]!=0)), names=rep('+', nrow(fit2)))

# MD plot (log fold change vs mean log expression)
# highlight statistically significant (p-adj < 0.05) probes
plotMD(fit2, column=ct, status=dT[,ct], legend=F, pch=20, cex=1)
abline(h=0)

# General expression data analysis
ex <- exprs(gset)

# box-and-whisker plot
ord <- order(gs)  # order samples by group
palette(c("#1B9E77", "#7570B3", "#E7298A", "#E6AB02", "#D95F02",
          "#66A61E", "#A6761D", "#B32424", "#B324B3", "#666666"))
par(mar=c(7,4,2,1))
title <- paste ("GSE114944", "/", annotation(gset), sep ="")
boxplot(ex[,ord], boxwex=0.6, notch=T, main=title, outline=FALSE, las=2, col=gs[ord])
legend("topleft", groups, fill=palette(), bty="n")

# expression value distribution
par(mar=c(4,4,2,1))
title <- paste ("GSE114944", "/", annotation(gset), " value distribution", sep ="")
plotDensities(ex, group=gs, main=title, legend ="topright")

# UMAP plot (dimensionality reduction)
ex <- na.omit(ex) # eliminate rows with NAs
ex <- ex[!duplicated(ex), ]  # remove duplicates
ump <- umap(t(ex), n_neighbors = 7, random_state = 123)
par(mar=c(3,3,2,6), xpd=TRUE)
plot(ump$layout, main="UMAP plot, nbrs=7", xlab="", ylab="", col=gs, pch=20, cex=1.5)
legend("topright", inset=c(-0.15,0), legend=levels(gs), pch=20,
col=1:nlevels(gs), title="Group", pt.cex=1.5)
library("maptools")  # point labels without overlaps
pointLabel(ump$layout, labels = rownames(ump$layout), method="SANN", cex=0.6)

# mean-variance trend, helps to see if precision weights are needed
plotSA(fit2, main="Mean variance trend, GSE114944")
```

## GEO2R Results

### Volcano plots

*Double positive vs. double negative*
![alt text](https://github.com/liuchen37/Pics/blob/main/114944_DPvsDN.jpg?raw=true)

| Gene Symbol | Gene title | Log2(fold change) | -Log10(*p*-value) |
| -- | -- | -- | -- |
| **Upregulation** |
| *CXCL13* | chemokine (C-X-C motif) ligand 13 | 6.45 | 6.405 
| *MYO7A* | myosin VIIA 4 | 5.095 | 4.958 
| *KIR2DL4* | killer cell immunoglobulin-like receptor, two domains, long cytoplasmic tail 4 | 4.528 | 5.237 
| *ENTPD1* | ectonucleoside triphosphate diphosphohydrolase 1 |3.914 | 6.117 
| *ETV1* |ets variant 1 | 3.796 | 4.695 
| **Downregulation** |
| *KLF3* | Kruppel-like factor 3 (basic) | -4.875 | 6.717 
| *PTPRM* | protein tyrosine phosphatase, receptor type, M |-4.485 | 6.535
| *PLAC8* | placenta specific 8 | -4.472 | 6.106
| *IL7R* |interleukin 7 receptor |-4.369 | 4.212
| *S1PR1* | sphingosine-1-phosphate receptor 1 | -4.129 | 4.406

*Double negative vs CD103 single positive*
![alt text](https://github.com/liuchen37/Pics/blob/main/114944_DNvsCD103SP.jpg?raw=true)

No significant genes.

*CD103 single positive vs double positive*
![alt text](https://github.com/liuchen37/Pics/blob/main/114944_CD103SPvsDP.jpg?raw=true)

| Gene Symbol | Gene title | Log2(fold change) | -Log10(*p*-value) |
| -- | -- | -- | -- |
| **Upregulation** |
|PTPRM | protein tyrosine phosphatase, receptor type, M | 3.376 | 4.995
| KLF3 | Kruppel-like factor 3 (basic) | 2.365 | 5.714
| **Downregulation** |
| KIR2DL4 | killer cell immunoglobulin-like receptor, two domains, long cytoplasmic tail, 4 | -5.251 | 6.035
| ENTPD1 | ectonucleoside triphosphate diphosphohydrolase 1 | -4.695 | 5.778
| KRT86 | keratin 86, type II | -3.295 | 5.667
| ENTPD1 | ectonucleoside triphosphate diphosphohydrolase 1 | -3.204 | 5.204
| HECTD2 | HECT domain containing E3 ubiquitin protein ligase 2 | -2.676 | 6.398


## Quality control

T-statistic quantile-quantile plot

![alt text](https://github.com/liuchen37/Pics/blob/main/114944_tqq_QC.jpg?raw=true)

```r
file.choose()

setwd("directory")

library(affy)
library(affyPLM)
library(RColorBrewer)

data <- ReadAffy()
Pset <- fitPLM(data)

# Chip-pseudo plot
image(data[,1])

# RLE plot utilising the "Set3" palette of RColorBrewer
colors <- brewer.pal(12, "Set3")
Mbox(Pset, col = colors, main = "RLE", las = 3)
```

![alt text](https://github.com/liuchen37/Pics/blob/main/gs_GSM3160643.png?raw=true)

![alt text](https://github.com/liuchen37/Pics/blob/main/RLE_GSE114944.png?raw=true)

## Raw data processing

Classify raw data (in .cel and .chp) locally into three subgroups: **DP, DN and CD103SP**.

```r
rm(list = ls())

library(affyPLM)
library(affy)
library(dplyr)

# process DP raw data
# load .cel file
file.choose()
setwd("directory")
DP <- ReadAffy()
sampleNames(DP)

# process with RMA method and output
dp.rma <-rma(DP)
dp.exp <- exprs(dp.rma)
probeid <- rownames(dp.exp)
dp.exp <- cbind(probeid, dp.exp)
write.table(dp.exp, file = "dp.txt", sep = '\t', quote = F, row.names = F)

# process DN raw data
# load .cel file
file.choose()
setwd("directory")
DN <- ReadAffy()
sampleNames(DN)

# process with RMA method and output
dn.rma <-rma(DN)
dn.exp <- exprs(dn.rma)
probeid <- rownames(dn.exp)
dn.exp <- cbind(probeid, dn.exp)
write.table(dn.exp, file = "dn.txt", sep = '\t', quote = F, row.names = F)

# process CD103SP raw data
# load .cel file
file.choose()
setwd("directory")
SP <- ReadAffy()
sampleNames(SP)

# process with RMA method and output
sp.rma <-rma(SP)
sp.exp <- exprs(sp.rma)
probeid <- rownames(sp.exp)
sp.exp <- cbind(probeid, sp.exp)
write.table(sp.exp, file = "sp.txt", sep = '\t', quote = F, row.names = F)

# merge tables
file.choose()
setwd("directory")
DN_expres <- read.table("dn.txt", header = T, sep = '\t')
DP_expres <- read.table("dp.txt", header = T, sep = '\t')
SP_expres <- read.table("sp.txt", header = T, sep = '\t')

Expression <- merge(DN_expres, DP_expres) %>%
  merge(SP_expres)

write.table(Expression, file = "expression.txt", sep = '\t', quote = F, row.names = F)
```

By doing the above steps, the following table consists of three groups of gene expression data was created:

![alt text](https://github.com/liuchen37/Pics/blob/main/expression114944.jpg?raw=true)

## Differential expression determination

### §1. Match probe ID with gene ID and symbol

The original GPL .txt file might be too large for R to process, therefore, remove excess columns in the file before processing, keep only probe ID, gene symbol and gene ID.

![alt text](https://github.com/liuchen37/Pics/blob/main/114944probe.jpg?raw=true)

```r
# Convert gene ID to Gene symbol
rm(list =ls())

file.choose()
setwd("directory")

# load expression file
expression <- read.table("expression.txt", header = T, sep = "\t", row.names = 1)

# load probe information
probe_info <- read.table("GPL15207-17536.txt", header = T, sep = "\t")
probe_name <- rownames(expression)

# match probeid with gene name
loc <- match(probe_info[,1], probe_name)

# as a filter, choose probes only have eligible expression
expression <- expression[loc,]
# in this case, 100 observations (probes) did not show any expression therefore being excluded

# match gprobe id with gene id (column 3)
raw_geneid <- as.numeric(as.matrix(probe_info[,3]))
# the "NA" warning prompted as there were invalid values created above

# extract probes that have geneid and output their expression and exclude "NA"
index <- which(!is.na(raw_geneid))
geneid <- raw_geneid[index]
exp_matrix <- expression[index,]
geneidfactor <- factor(geneid)

# if multiple probe detecting one gene, calculate their mean value
expression_m <- apply(exp_matrix, 2, function (x) tapply(x, geneidfactor, mean))

# set geneid as row names
rownames(expression_m) <- levels(geneidfactor)
geneid <- rownames(expression_m)
expression_m_2 <- cbind(geneid, expression_m)
write.table(expression_m_2, file = "TIL expression.txt", sep = '\t', quote= F, row.names = F)

# convert geneid to gene symbol
loc <- match(rownames(expression_m), probe_info[,3])
rownames(expression_m) = probe_info[loc,2]
genesymbol <- rownames(expression_m)
expression_m_3 <- cbind(genesymbol, expression_m)
write.table(expression_m_3, file = "TIL gene symbol.txt", sep = '\t', quote= F, row.names = F)
```

By completing the above steps, a table consists of expression value and gene ID, and a table consists of expression value and gene symbol would be created.

*Expression and Gene ID on probe:*

![alt text](https://github.com/liuchen37/Pics/blob/main/114944_with%20probe%20gene%20id.jpg?raw=true)

*Expression and Gene symbol:*

![alt text](https://github.com/liuchen37/Pics/blob/main/114944%20_with%20gene%20symbol.jpg?raw=true)

```r
# use K-nearest Neighbor (KNN) mathod to replace lost values for each probe
library(impute)

# load TIL expression with gene symbol
file.choose()
gene_exp_matrix <- read.table("TIL gene symbol.txt", header = T, sep = '\t', row.names = 1)
## Error in read.table("TIL gene symbol.txt", header = T, sep = "\t", row.names = 1) : 
##    duplicate 'row.names' are not allowed
```

This error is due to expressions (rows) failed to match a gene symbol with the probe id. Therefore the blank gene symbol was replaced by a date. Loacte them and delete the blank expressions could solve the problem.

![alt tex](https://github.com/liuchen37/Pics/blob/main/114944_probes%20does%20not%20have%20a%20gene%20symbol.jpg?raw=true)

```r
gene_exp_matrix <- read.table("TIL gene symbol.txt", header = T, sep = '\t', row.names = 1)
gene_exp_matrix <- as.matrix(gene_exp_matrix)

# apply KNN
imputed_gene_exp <- impute.knn(gene_exp_matrix, k = 10, rowmax = 0.5, colmax = 0.8, maxp = 3000, rng.seed = 362436069)
## Cluster size 18810 broken into 9175 9635 
## Cluster size 9175 broken into 5554 3621 
## Cluster size 5554 broken into 1936 3618 
## Done cluster 1936 
## Cluster size 3618 broken into 2170 1448 
## Done cluster 2170 
## Done cluster 1448 
## Done cluster 3618 
## Done cluster 5554 
## Cluster size 3621 broken into 1615 2006 
## Done cluster 1615 
## Done cluster 2006 
## Done cluster 3621 
## Done cluster 9175 
## Cluster size 9635 broken into 2713 6922 
## Done cluster 2713 
## Cluster size 6922 broken into 3321 3601 
## Cluster size 3321 broken into 1869 1452 
## Done cluster 1869 
## Done cluster 1452 
## Done cluster 3321 
## Cluster size 3601 broken into 2381 1220 
## Done cluster 2381 
## ## Done cluster 1220 
## Done cluster 3601 
## Done cluster 6922 
## Done cluster 9635

# output results
GeneExp <- imputed_gene_exp$data
genesymbol <- rownames(GeneExp)
GeneExp <- cbind(genesymbol, GeneExp)
write.table(GeneExp, file = "TIL gene expression final.txt", sep = '\t', quote = F, row.names = F)
```

By completing the above steps, a table of gene symbol and its epxression would be created, the blank expression had been filled in by the KNN method.

![alt text](https://github.com/liuchen37/Pics/blob/main/114944_gene%20expression.jpg?raw=true)

### §2. Differential expression determination

```r
library(limma)
rt <- read.table("TIL gene expression final.txt", header = T, sep = '\t', row.names = "genesymbol")

# classify samples as previous steps, specifically note that sample number must match their sub group in a correct order, in this case is 5xDN, 5xDP and 5xCD103SP
class <- c(rep("DN", 5), rep("DP", 5), rep("CD103SP", 5))
design <- model.matrix(~factor(class))
colnames(design) <- c("DN", "DP", "CD103SP")

# calculate mean values and variances of the three groups and test via Bayes inference
fit <- lmFit(rt, design)
fit2 <- eBayes(fit)

# output all genes, regardless of its variance
allDiff = topTable(fit2, adjust = 'fdr', coef = 2, number = 200000)
write.table(allDiff, file = "differential.xls", sep = '\t', quote = F)

# determine differential expression genes, based on fold change
Diff <- allDiff[with(allDiff, ((logFC > 1 | logFC < (-1)) & adj.P.Val < 0.05)),]
write.table(Diff, file = "diffExp.xls", sep = '\t', quote = F)

# co-expression
diffExpLevel <- rt[rownames(Diff),]
qvalue = allDiff[rownames(Diff),]$adj.P.Val
diffExpQvalue = cbind(qvalue, diffExpLevel)
write.table(diffExpQvalue, file = "diffExpLevel.xls", sep = '\t', quote = F)
```

By completing the above steps, a table containing diffrential expression information would be created:

![alt text](https://github.com/liuchen37/Pics/blob/main/14944_differential.jpg?raw=true)

### §3. Producing heatmap
